Xamarin.Forms.Core/IndicatorView.cs

```diff
@@ -0,0 +1,256 @@
+		{
+			get => (int)GetValue(CountProperty);
+			set => SetValue(CountProperty, value);
+		}
+
+		public int MaximumVisibleCount
+		{
+			get => (int)GetValue(MaximumVisibleCountProperty);
+			set => SetValue(MaximumVisibleCountProperty, value);
+		}
+
+		public DataTemplate IndicatorTemplate
+		{
+			get => (DataTemplate)GetValue(IndicatorTemplateProperty);
+			set => SetValue(IndicatorTemplateProperty, value);
+		}
+
+		public bool HidesForSingleIndicator
+		{
+			get => (bool)GetValue(HidesForSingleIndicatorProperty);
+			set => SetValue(HidesForSingleIndicatorProperty, value);
+		}
+
+		public Color IndicatorColor
+		{
+			get => (Color)GetValue(IndicatorColorProperty);
+			set => SetValue(IndicatorColorProperty, value);
+		}
+
+		public Color SelectedIndicatorColor
+		{
+			get => (Color)GetValue(SelectedIndicatorColorProperty);
+			set => SetValue(SelectedIndicatorColorProperty, value);
+		}
+
+		public double IndicatorSize
+		{
+			get => (double)GetValue(IndicatorSizeProperty);
+			set => SetValue(IndicatorSizeProperty, value);
+		}
+
+		public IEnumerable ItemsSource
+		{
+			get => (IEnumerable)GetValue(ItemsSourceProperty);
+			set => SetValue(ItemsSourceProperty, value);
+		}
+
+		void ResetIndicatorStyles()
+		{
+			try
+			{
+				BatchBegin();
+				ResetIndicatorStylesNonBatch();
+			}
+			finally
+			{
+				BatchCommit();
+			}
+		}
+
+		void ResetIndicatorCount(int oldCount)
+		{
+			try
+			{
+				BatchBegin();
+				if (oldCount < 0)
+				{
+					oldCount = 0;
+				}
+
+				if (oldCount > Count)
+				{
+					RemoveRedundantIndicatorItems();
+					return;
+				}
+
+				AddExtraIndicatorItems();
+			}
+			finally
+			{
+				ResetIndicatorStylesNonBatch();
+				BatchCommit();
+			}
+		}
+
+		void ResetIndicatorStylesNonBatch()
+		{
+			foreach (var child in Items)
+			{
+				ApplyColor(child as View);
+			}
+
+			IndicatorLayout.IsVisible = Count > 1 || !HidesForSingleIndicator;
+		}
+
+		void ResetItemsSource(IEnumerable oldItemsSource)
+		{
+			if (oldItemsSource is INotifyCollectionChanged oldObservableCollection)
+			{
+				oldObservableCollection.CollectionChanged -= OnCollectionChanged;
+			}
+
+			if (ItemsSource is INotifyCollectionChanged observableCollection)
+			{
+				observableCollection.CollectionChanged += OnCollectionChanged;
+			}
+
+			OnCollectionChanged(ItemsSource, new NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction.Reset));
+		}
+
+		void AddExtraIndicatorItems()
+		{
+			var oldCount = Items.Count;
+			for (var i = 0; i < Count - oldCount && i < MaximumVisibleCount - oldCount; ++i)
+			{
+				var size = IndicatorSize > 0 ? IndicatorSize : 10;
+				var indicator = IndicatorTemplate?.CreateContent() as View ?? new Frame
+				{
+					Padding = 0,
+					HasShadow = false,
+					BorderColor = Color.Transparent,
+					VerticalOptions = LayoutOptions.Center,
+					HorizontalOptions = LayoutOptions.Center,
+					WidthRequest = size,
+					HeightRequest = size,
+					CornerRadius = (float)size / 2
```

Comment:
To get a perfect circular `Frame`, `CornerRadius` needs to be exactly half the size of frame width and height. Since `IndicatorSize` is a double and you're doing a division here while casting the result to float, I believe you might end up with rounding issues in some cases and unexpected "borders". @davidortinau Is the Drawing Spec still on? Wish we could get our own primitive shapes.

Reply:
hmmmm

Reply:
Again, this issue could be solved in native renderers to leverage native drawing APIs. Forms is slowing us down here a bit.